{
 "metadata": {
  "name": "",
  "signature": "sha256:d5611e5a77bade40451d28f2da0fd35f16c015b62bd1e7901573f37f0c328e37"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "heading",
     "level": 1,
     "metadata": {},
     "source": [
      "Introduction"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "The Wold Lab started processing our samples before the NIST spikes were available, and so to allow historical comparison we continue adding our spikes into our libraries. We did have two different variations of the concentrations of the Caltech spikes.\n",
      "\n",
      "The Caltech spikes sequence is available as <a href=\"https://www.encodedcc.org/ENCFF001RTO\">ENCFF001RTO</a>. The NIST spikes are available as <a href=\"https://www.encodedcc.org/ENCFF001RTP\">ENCFF001RTP</a>."
     ]
    },
    {
     "cell_type": "heading",
     "level": 2,
     "metadata": {},
     "source": [
      "Setup"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "We need to make sure that we use the encode approved versions of the mapping tools.\n",
      "\n",
      "Use this <a href=\"https://docs.google.com/spreadsheets/d/14j1Jl53_VBXekqRqoU8qm7B2YOthMDFBP4QdJroZaX8/edit#gid=0\">spreadsheet</a> as the reference.\n",
      "\n",
      "  * bowtie 2 2.1.0 /woldlab/castor/data00/proj/genome/programs/bowtie-2.1.0/bowtie2\n",
      "  * bowtie 1 0.12.9 /woldlab/castor/data00/proj/genome/programs/bowtie-0.12.9/bowtie\n",
      "  * tophat 1.0.8"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "Before running this notebook make sure you've done the following (or its equivalent) as defined in the Makefile\n",
      "\n",
      "* <code>mkdir indexes</code>\n",
      "* <code>mkdir indexes2</code>\n",
      "* <code>make</code>\n",
      "\n",
      "Make all is doing 3 things:\n",
      "* (<code>make spikes</code>) building fasta files of the NIST + Caltech spikes containing a sliding window of all the spike sequence at several read lengths. \n",
      "* (<code>make indexes</code>) builds bowtie 1 indexes of the NIST + Caltech spikes by themsevels and with the hg19 male and female genomes.\n",
      "* (<code>make indexes2</code>) builds bowtie 2 indexes of above"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "from __future__ import print_function"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 1
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "import collections\n",
      "import os\n",
      "import re\n",
      "import pandas"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 2
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "read_lengths = [10, 36, 75, 100]"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 3
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "mismatches = [0, 1, 2, 3]"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 4
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "bowtie1_indexes = [\n",
      "  'indexes/spikes', \n",
      "  'indexes/ENCFF001RGR+spikes', \n",
      "  'indexes/ENCFF001RGS+spikes',\n",
      "  'gencode.v19/ENCFF001RGR+spikes',\n",
      "  'gencode.v19/ENCFF001RGS+spikes',\n",
      "  ]\n",
      "\n",
      "bowtie2_indexes = [\n",
      "  'indexes2/spikes', \n",
      "  'indexes2/ENCFF001RGR+spikes', \n",
      "  'indexes2/ENCFF001RGS+spikes',\n",
      "  'gencode.v19.bt2/ENCFF001RGR+spikes',\n",
      "  'gencode.v19.bt2/ENCFF001RGS+spikes',\n",
      "  ]"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 5
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "cache = pandas.HDFStore('cache')"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 6
    },
    {
     "cell_type": "heading",
     "level": 2,
     "metadata": {},
     "source": [
      "Analyze spikes with bowtie 1"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "! /woldlab/castor/data00/proj/genome/programs/bowtie-0.12.9/bowtie --version"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "/woldlab/castor/data00/proj/genome/programs/bowtie-0.12.9/bowtie version 0.12.9\r\n",
        "64-bit\r\n",
        "Built on igm1\r\n",
        "Sun Dec 16 14:36:32 EST 2012\r\n",
        "Compiler: gcc version 4.1.2 20080704 (Red Hat 4.1.2-50)\r\n",
        "Options: -O3 -m64  -Wl,--hash-style=both  \r\n",
        "Sizeof {int, long, long long, void*, size_t, off_t}: {4, 8, 8, 8, 8, 8}\r\n"
       ]
      }
     ],
     "prompt_number": 7
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def parse_bowtie1(result):\n",
      "    parsed = {'total': 0,\n",
      "              'hits': 0,\n",
      "              'failed': 0,\n",
      "              'multi': 0}\n",
      "    \n",
      "    terms = [ ('reads processed: ([\\d]+)', 'total'),\n",
      "              ('one reported alignment: ([\\d]+)', 'hits'),\n",
      "              ('failed to align: ([\\d]+)', 'failed'),\n",
      "              ('alignments suppressed due to -m: ([\\d]+)', 'multi')]\n",
      "    for line in result:\n",
      "        for query, variable in terms:\n",
      "            match = re.search(query, line)\n",
      "            if match:\n",
      "                parsed[variable] = int(match.group(1))\n",
      "    return parsed"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 8
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      " def bowtie(index_pathname, mismatches, read_length):\n",
      "    \"\"\"Using some ipython features run bowtie from python\n",
      "    \n",
      "    -v specifies the allowed number of mismatches in a range 0-3.\n",
      "    -k (return upto N matches)\n",
      "    -m (suppress reads with more than ths number of hits)\n",
      "    \n",
      "    Using -m 1 should require all reads to be unique.\n",
      "    \"\"\"\n",
      "    path, index_name = os.path.split(index_pathname)\n",
      "    result = ! /woldlab/castor/data00/proj/genome/programs/bowtie-0.12.9/bowtie \\\n",
      "                      {index_pathname}  -v {mismatches} -m 1 --best --strata -f \\\n",
      "                      spikes-{read_length}.fa \\\n",
      "                      --un {mismatches}-mm/unmapped-{index_name}-{read_length}.fa \\\n",
      "                      --max {mismatches}-mm/repeat-{index_name}-{read_length}.fa \\\n",
      "                      {mismatches}-mm/hits-{index_name}-{read_length}.txt\n",
      "    return parse_bowtie1(result)\n",
      "                "
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 9
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def process_bowtie1(indexes, force=False):\n",
      "    # uncomment for testing.\n",
      "    #indexes = ['spikes']\n",
      "    cache_name = 'bowtie1'\n",
      "    if cache_name in cache and not force:\n",
      "        return cache[cache_name]\n",
      "    processed = []\n",
      "    for index in indexes:\n",
      "        for mismatch in mismatches:\n",
      "            for read_length in read_lengths:\n",
      "                properties = {\n",
      "                   'index': index,\n",
      "                   'mismatch': mismatch,\n",
      "                   'read_length': read_length}\n",
      "                result = bowtie(index, mismatch, read_length)\n",
      "                properties.update(result)\n",
      "                processed.append(properties)\n",
      "    df =  pandas.DataFrame(processed, columns=['index', 'mismatch', 'read_length', 'total', 'hits', 'multi', 'failed', ])\n",
      "    cache[cache_name] = df\n",
      "    return df"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 10
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "bowtie1_results = process_bowtie1(bowtie1_indexes)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 11
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "Bowtie 1 Genomic"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "bowtie1_results[:48]"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "html": [
        "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
        "<table border=\"1\" class=\"dataframe\">\n",
        "  <thead>\n",
        "    <tr style=\"text-align: right;\">\n",
        "      <th></th>\n",
        "      <th>index</th>\n",
        "      <th>mismatch</th>\n",
        "      <th>read_length</th>\n",
        "      <th>total</th>\n",
        "      <th>hits</th>\n",
        "      <th>multi</th>\n",
        "      <th>failed</th>\n",
        "    </tr>\n",
        "  </thead>\n",
        "  <tbody>\n",
        "    <tr>\n",
        "      <th>0 </th>\n",
        "      <td>             indexes/spikes</td>\n",
        "      <td> 0</td>\n",
        "      <td>  10</td>\n",
        "      <td> 111807</td>\n",
        "      <td>  85527</td>\n",
        "      <td>  26276</td>\n",
        "      <td>   4</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>1 </th>\n",
        "      <td>             indexes/spikes</td>\n",
        "      <td> 0</td>\n",
        "      <td>  36</td>\n",
        "      <td> 109103</td>\n",
        "      <td> 109067</td>\n",
        "      <td>      0</td>\n",
        "      <td>  36</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>2 </th>\n",
        "      <td>             indexes/spikes</td>\n",
        "      <td> 0</td>\n",
        "      <td>  75</td>\n",
        "      <td> 105047</td>\n",
        "      <td> 104972</td>\n",
        "      <td>      0</td>\n",
        "      <td>  75</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>3 </th>\n",
        "      <td>             indexes/spikes</td>\n",
        "      <td> 0</td>\n",
        "      <td> 100</td>\n",
        "      <td> 102447</td>\n",
        "      <td> 102347</td>\n",
        "      <td>      0</td>\n",
        "      <td> 100</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>4 </th>\n",
        "      <td>             indexes/spikes</td>\n",
        "      <td> 1</td>\n",
        "      <td>  10</td>\n",
        "      <td> 111807</td>\n",
        "      <td>  85527</td>\n",
        "      <td>  26280</td>\n",
        "      <td>   0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>5 </th>\n",
        "      <td>             indexes/spikes</td>\n",
        "      <td> 1</td>\n",
        "      <td>  36</td>\n",
        "      <td> 109103</td>\n",
        "      <td> 109067</td>\n",
        "      <td>      0</td>\n",
        "      <td>  36</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>6 </th>\n",
        "      <td>             indexes/spikes</td>\n",
        "      <td> 1</td>\n",
        "      <td>  75</td>\n",
        "      <td> 105047</td>\n",
        "      <td> 104972</td>\n",
        "      <td>      0</td>\n",
        "      <td>  75</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>7 </th>\n",
        "      <td>             indexes/spikes</td>\n",
        "      <td> 1</td>\n",
        "      <td> 100</td>\n",
        "      <td> 102447</td>\n",
        "      <td> 102347</td>\n",
        "      <td>      0</td>\n",
        "      <td> 100</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>8 </th>\n",
        "      <td>             indexes/spikes</td>\n",
        "      <td> 2</td>\n",
        "      <td>  10</td>\n",
        "      <td> 111807</td>\n",
        "      <td>  85527</td>\n",
        "      <td>  26280</td>\n",
        "      <td>   0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>9 </th>\n",
        "      <td>             indexes/spikes</td>\n",
        "      <td> 2</td>\n",
        "      <td>  36</td>\n",
        "      <td> 109103</td>\n",
        "      <td> 109067</td>\n",
        "      <td>      0</td>\n",
        "      <td>  36</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>10</th>\n",
        "      <td>             indexes/spikes</td>\n",
        "      <td> 2</td>\n",
        "      <td>  75</td>\n",
        "      <td> 105047</td>\n",
        "      <td> 104972</td>\n",
        "      <td>      0</td>\n",
        "      <td>  75</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>11</th>\n",
        "      <td>             indexes/spikes</td>\n",
        "      <td> 2</td>\n",
        "      <td> 100</td>\n",
        "      <td> 102447</td>\n",
        "      <td> 102347</td>\n",
        "      <td>      0</td>\n",
        "      <td> 100</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>12</th>\n",
        "      <td>             indexes/spikes</td>\n",
        "      <td> 3</td>\n",
        "      <td>  10</td>\n",
        "      <td> 111807</td>\n",
        "      <td>  85527</td>\n",
        "      <td>  26280</td>\n",
        "      <td>   0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>13</th>\n",
        "      <td>             indexes/spikes</td>\n",
        "      <td> 3</td>\n",
        "      <td>  36</td>\n",
        "      <td> 109103</td>\n",
        "      <td> 109067</td>\n",
        "      <td>      0</td>\n",
        "      <td>  36</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>14</th>\n",
        "      <td>             indexes/spikes</td>\n",
        "      <td> 3</td>\n",
        "      <td>  75</td>\n",
        "      <td> 105047</td>\n",
        "      <td> 104972</td>\n",
        "      <td>      0</td>\n",
        "      <td>  75</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>15</th>\n",
        "      <td>             indexes/spikes</td>\n",
        "      <td> 3</td>\n",
        "      <td> 100</td>\n",
        "      <td> 102447</td>\n",
        "      <td> 102347</td>\n",
        "      <td>      0</td>\n",
        "      <td> 100</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>16</th>\n",
        "      <td> indexes/ENCFF001RGR+spikes</td>\n",
        "      <td> 0</td>\n",
        "      <td>  10</td>\n",
        "      <td> 111807</td>\n",
        "      <td>      0</td>\n",
        "      <td> 111807</td>\n",
        "      <td>   0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>17</th>\n",
        "      <td> indexes/ENCFF001RGR+spikes</td>\n",
        "      <td> 0</td>\n",
        "      <td>  36</td>\n",
        "      <td> 109103</td>\n",
        "      <td> 109067</td>\n",
        "      <td>      0</td>\n",
        "      <td>  36</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>18</th>\n",
        "      <td> indexes/ENCFF001RGR+spikes</td>\n",
        "      <td> 0</td>\n",
        "      <td>  75</td>\n",
        "      <td> 105047</td>\n",
        "      <td> 104972</td>\n",
        "      <td>      0</td>\n",
        "      <td>  75</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>19</th>\n",
        "      <td> indexes/ENCFF001RGR+spikes</td>\n",
        "      <td> 0</td>\n",
        "      <td> 100</td>\n",
        "      <td> 102447</td>\n",
        "      <td> 102347</td>\n",
        "      <td>      0</td>\n",
        "      <td> 100</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>20</th>\n",
        "      <td> indexes/ENCFF001RGR+spikes</td>\n",
        "      <td> 1</td>\n",
        "      <td>  10</td>\n",
        "      <td> 111807</td>\n",
        "      <td>      0</td>\n",
        "      <td> 111807</td>\n",
        "      <td>   0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>21</th>\n",
        "      <td> indexes/ENCFF001RGR+spikes</td>\n",
        "      <td> 1</td>\n",
        "      <td>  36</td>\n",
        "      <td> 109103</td>\n",
        "      <td> 109067</td>\n",
        "      <td>      0</td>\n",
        "      <td>  36</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>22</th>\n",
        "      <td> indexes/ENCFF001RGR+spikes</td>\n",
        "      <td> 1</td>\n",
        "      <td>  75</td>\n",
        "      <td> 105047</td>\n",
        "      <td> 104972</td>\n",
        "      <td>      0</td>\n",
        "      <td>  75</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>23</th>\n",
        "      <td> indexes/ENCFF001RGR+spikes</td>\n",
        "      <td> 1</td>\n",
        "      <td> 100</td>\n",
        "      <td> 102447</td>\n",
        "      <td> 102347</td>\n",
        "      <td>      0</td>\n",
        "      <td> 100</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>24</th>\n",
        "      <td> indexes/ENCFF001RGR+spikes</td>\n",
        "      <td> 2</td>\n",
        "      <td>  10</td>\n",
        "      <td> 111807</td>\n",
        "      <td>      0</td>\n",
        "      <td> 111807</td>\n",
        "      <td>   0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>25</th>\n",
        "      <td> indexes/ENCFF001RGR+spikes</td>\n",
        "      <td> 2</td>\n",
        "      <td>  36</td>\n",
        "      <td> 109103</td>\n",
        "      <td> 109067</td>\n",
        "      <td>      0</td>\n",
        "      <td>  36</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>26</th>\n",
        "      <td> indexes/ENCFF001RGR+spikes</td>\n",
        "      <td> 2</td>\n",
        "      <td>  75</td>\n",
        "      <td> 105047</td>\n",
        "      <td> 104972</td>\n",
        "      <td>      0</td>\n",
        "      <td>  75</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>27</th>\n",
        "      <td> indexes/ENCFF001RGR+spikes</td>\n",
        "      <td> 2</td>\n",
        "      <td> 100</td>\n",
        "      <td> 102447</td>\n",
        "      <td> 102347</td>\n",
        "      <td>      0</td>\n",
        "      <td> 100</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>28</th>\n",
        "      <td> indexes/ENCFF001RGR+spikes</td>\n",
        "      <td> 3</td>\n",
        "      <td>  10</td>\n",
        "      <td> 111807</td>\n",
        "      <td>      0</td>\n",
        "      <td> 111807</td>\n",
        "      <td>   0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>29</th>\n",
        "      <td> indexes/ENCFF001RGR+spikes</td>\n",
        "      <td> 3</td>\n",
        "      <td>  36</td>\n",
        "      <td> 109103</td>\n",
        "      <td> 109067</td>\n",
        "      <td>      0</td>\n",
        "      <td>  36</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>30</th>\n",
        "      <td> indexes/ENCFF001RGR+spikes</td>\n",
        "      <td> 3</td>\n",
        "      <td>  75</td>\n",
        "      <td> 105047</td>\n",
        "      <td> 104972</td>\n",
        "      <td>      0</td>\n",
        "      <td>  75</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>31</th>\n",
        "      <td> indexes/ENCFF001RGR+spikes</td>\n",
        "      <td> 3</td>\n",
        "      <td> 100</td>\n",
        "      <td> 102447</td>\n",
        "      <td> 102347</td>\n",
        "      <td>      0</td>\n",
        "      <td> 100</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>32</th>\n",
        "      <td> indexes/ENCFF001RGS+spikes</td>\n",
        "      <td> 0</td>\n",
        "      <td>  10</td>\n",
        "      <td> 111807</td>\n",
        "      <td>      0</td>\n",
        "      <td> 111807</td>\n",
        "      <td>   0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>33</th>\n",
        "      <td> indexes/ENCFF001RGS+spikes</td>\n",
        "      <td> 0</td>\n",
        "      <td>  36</td>\n",
        "      <td> 109103</td>\n",
        "      <td> 109067</td>\n",
        "      <td>      0</td>\n",
        "      <td>  36</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>34</th>\n",
        "      <td> indexes/ENCFF001RGS+spikes</td>\n",
        "      <td> 0</td>\n",
        "      <td>  75</td>\n",
        "      <td> 105047</td>\n",
        "      <td> 104972</td>\n",
        "      <td>      0</td>\n",
        "      <td>  75</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>35</th>\n",
        "      <td> indexes/ENCFF001RGS+spikes</td>\n",
        "      <td> 0</td>\n",
        "      <td> 100</td>\n",
        "      <td> 102447</td>\n",
        "      <td> 102347</td>\n",
        "      <td>      0</td>\n",
        "      <td> 100</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>36</th>\n",
        "      <td> indexes/ENCFF001RGS+spikes</td>\n",
        "      <td> 1</td>\n",
        "      <td>  10</td>\n",
        "      <td> 111807</td>\n",
        "      <td>      0</td>\n",
        "      <td> 111807</td>\n",
        "      <td>   0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>37</th>\n",
        "      <td> indexes/ENCFF001RGS+spikes</td>\n",
        "      <td> 1</td>\n",
        "      <td>  36</td>\n",
        "      <td> 109103</td>\n",
        "      <td> 109067</td>\n",
        "      <td>      0</td>\n",
        "      <td>  36</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>38</th>\n",
        "      <td> indexes/ENCFF001RGS+spikes</td>\n",
        "      <td> 1</td>\n",
        "      <td>  75</td>\n",
        "      <td> 105047</td>\n",
        "      <td> 104972</td>\n",
        "      <td>      0</td>\n",
        "      <td>  75</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>39</th>\n",
        "      <td> indexes/ENCFF001RGS+spikes</td>\n",
        "      <td> 1</td>\n",
        "      <td> 100</td>\n",
        "      <td> 102447</td>\n",
        "      <td> 102347</td>\n",
        "      <td>      0</td>\n",
        "      <td> 100</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>40</th>\n",
        "      <td> indexes/ENCFF001RGS+spikes</td>\n",
        "      <td> 2</td>\n",
        "      <td>  10</td>\n",
        "      <td> 111807</td>\n",
        "      <td>      0</td>\n",
        "      <td> 111807</td>\n",
        "      <td>   0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>41</th>\n",
        "      <td> indexes/ENCFF001RGS+spikes</td>\n",
        "      <td> 2</td>\n",
        "      <td>  36</td>\n",
        "      <td> 109103</td>\n",
        "      <td> 109067</td>\n",
        "      <td>      0</td>\n",
        "      <td>  36</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>42</th>\n",
        "      <td> indexes/ENCFF001RGS+spikes</td>\n",
        "      <td> 2</td>\n",
        "      <td>  75</td>\n",
        "      <td> 105047</td>\n",
        "      <td> 104972</td>\n",
        "      <td>      0</td>\n",
        "      <td>  75</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>43</th>\n",
        "      <td> indexes/ENCFF001RGS+spikes</td>\n",
        "      <td> 2</td>\n",
        "      <td> 100</td>\n",
        "      <td> 102447</td>\n",
        "      <td> 102347</td>\n",
        "      <td>      0</td>\n",
        "      <td> 100</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>44</th>\n",
        "      <td> indexes/ENCFF001RGS+spikes</td>\n",
        "      <td> 3</td>\n",
        "      <td>  10</td>\n",
        "      <td> 111807</td>\n",
        "      <td>      0</td>\n",
        "      <td> 111807</td>\n",
        "      <td>   0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>45</th>\n",
        "      <td> indexes/ENCFF001RGS+spikes</td>\n",
        "      <td> 3</td>\n",
        "      <td>  36</td>\n",
        "      <td> 109103</td>\n",
        "      <td> 109067</td>\n",
        "      <td>      0</td>\n",
        "      <td>  36</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>46</th>\n",
        "      <td> indexes/ENCFF001RGS+spikes</td>\n",
        "      <td> 3</td>\n",
        "      <td>  75</td>\n",
        "      <td> 105047</td>\n",
        "      <td> 104972</td>\n",
        "      <td>      0</td>\n",
        "      <td>  75</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>47</th>\n",
        "      <td> indexes/ENCFF001RGS+spikes</td>\n",
        "      <td> 3</td>\n",
        "      <td> 100</td>\n",
        "      <td> 102447</td>\n",
        "      <td> 102347</td>\n",
        "      <td>      0</td>\n",
        "      <td> 100</td>\n",
        "    </tr>\n",
        "  </tbody>\n",
        "</table>\n",
        "</div>"
       ],
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 12,
       "text": [
        "                         index  mismatch  read_length   total    hits   multi  \\\n",
        "0               indexes/spikes         0           10  111807   85527   26276   \n",
        "1               indexes/spikes         0           36  109103  109067       0   \n",
        "2               indexes/spikes         0           75  105047  104972       0   \n",
        "3               indexes/spikes         0          100  102447  102347       0   \n",
        "4               indexes/spikes         1           10  111807   85527   26280   \n",
        "5               indexes/spikes         1           36  109103  109067       0   \n",
        "6               indexes/spikes         1           75  105047  104972       0   \n",
        "7               indexes/spikes         1          100  102447  102347       0   \n",
        "8               indexes/spikes         2           10  111807   85527   26280   \n",
        "9               indexes/spikes         2           36  109103  109067       0   \n",
        "10              indexes/spikes         2           75  105047  104972       0   \n",
        "11              indexes/spikes         2          100  102447  102347       0   \n",
        "12              indexes/spikes         3           10  111807   85527   26280   \n",
        "13              indexes/spikes         3           36  109103  109067       0   \n",
        "14              indexes/spikes         3           75  105047  104972       0   \n",
        "15              indexes/spikes         3          100  102447  102347       0   \n",
        "16  indexes/ENCFF001RGR+spikes         0           10  111807       0  111807   \n",
        "17  indexes/ENCFF001RGR+spikes         0           36  109103  109067       0   \n",
        "18  indexes/ENCFF001RGR+spikes         0           75  105047  104972       0   \n",
        "19  indexes/ENCFF001RGR+spikes         0          100  102447  102347       0   \n",
        "20  indexes/ENCFF001RGR+spikes         1           10  111807       0  111807   \n",
        "21  indexes/ENCFF001RGR+spikes         1           36  109103  109067       0   \n",
        "22  indexes/ENCFF001RGR+spikes         1           75  105047  104972       0   \n",
        "23  indexes/ENCFF001RGR+spikes         1          100  102447  102347       0   \n",
        "24  indexes/ENCFF001RGR+spikes         2           10  111807       0  111807   \n",
        "25  indexes/ENCFF001RGR+spikes         2           36  109103  109067       0   \n",
        "26  indexes/ENCFF001RGR+spikes         2           75  105047  104972       0   \n",
        "27  indexes/ENCFF001RGR+spikes         2          100  102447  102347       0   \n",
        "28  indexes/ENCFF001RGR+spikes         3           10  111807       0  111807   \n",
        "29  indexes/ENCFF001RGR+spikes         3           36  109103  109067       0   \n",
        "30  indexes/ENCFF001RGR+spikes         3           75  105047  104972       0   \n",
        "31  indexes/ENCFF001RGR+spikes         3          100  102447  102347       0   \n",
        "32  indexes/ENCFF001RGS+spikes         0           10  111807       0  111807   \n",
        "33  indexes/ENCFF001RGS+spikes         0           36  109103  109067       0   \n",
        "34  indexes/ENCFF001RGS+spikes         0           75  105047  104972       0   \n",
        "35  indexes/ENCFF001RGS+spikes         0          100  102447  102347       0   \n",
        "36  indexes/ENCFF001RGS+spikes         1           10  111807       0  111807   \n",
        "37  indexes/ENCFF001RGS+spikes         1           36  109103  109067       0   \n",
        "38  indexes/ENCFF001RGS+spikes         1           75  105047  104972       0   \n",
        "39  indexes/ENCFF001RGS+spikes         1          100  102447  102347       0   \n",
        "40  indexes/ENCFF001RGS+spikes         2           10  111807       0  111807   \n",
        "41  indexes/ENCFF001RGS+spikes         2           36  109103  109067       0   \n",
        "42  indexes/ENCFF001RGS+spikes         2           75  105047  104972       0   \n",
        "43  indexes/ENCFF001RGS+spikes         2          100  102447  102347       0   \n",
        "44  indexes/ENCFF001RGS+spikes         3           10  111807       0  111807   \n",
        "45  indexes/ENCFF001RGS+spikes         3           36  109103  109067       0   \n",
        "46  indexes/ENCFF001RGS+spikes         3           75  105047  104972       0   \n",
        "47  indexes/ENCFF001RGS+spikes         3          100  102447  102347       0   \n",
        "\n",
        "    failed  \n",
        "0        4  \n",
        "1       36  \n",
        "2       75  \n",
        "3      100  \n",
        "4        0  \n",
        "5       36  \n",
        "6       75  \n",
        "7      100  \n",
        "8        0  \n",
        "9       36  \n",
        "10      75  \n",
        "11     100  \n",
        "12       0  \n",
        "13      36  \n",
        "14      75  \n",
        "15     100  \n",
        "16       0  \n",
        "17      36  \n",
        "18      75  \n",
        "19     100  \n",
        "20       0  \n",
        "21      36  \n",
        "22      75  \n",
        "23     100  \n",
        "24       0  \n",
        "25      36  \n",
        "26      75  \n",
        "27     100  \n",
        "28       0  \n",
        "29      36  \n",
        "30      75  \n",
        "31     100  \n",
        "32       0  \n",
        "33      36  \n",
        "34      75  \n",
        "35     100  \n",
        "36       0  \n",
        "37      36  \n",
        "38      75  \n",
        "39     100  \n",
        "40       0  \n",
        "41      36  \n",
        "42      75  \n",
        "43     100  \n",
        "44       0  \n",
        "45      36  \n",
        "46      75  \n",
        "47     100  "
       ]
      }
     ],
     "prompt_number": 12
    },
    {
     "cell_type": "heading",
     "level": 2,
     "metadata": {},
     "source": [
      "Bowtie 1 transcriptome"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "bowtie1_results[48:]"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "html": [
        "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
        "<table border=\"1\" class=\"dataframe\">\n",
        "  <thead>\n",
        "    <tr style=\"text-align: right;\">\n",
        "      <th></th>\n",
        "      <th>index</th>\n",
        "      <th>mismatch</th>\n",
        "      <th>read_length</th>\n",
        "      <th>total</th>\n",
        "      <th>hits</th>\n",
        "      <th>multi</th>\n",
        "      <th>failed</th>\n",
        "    </tr>\n",
        "  </thead>\n",
        "  <tbody>\n",
        "    <tr>\n",
        "      <th>48</th>\n",
        "      <td> gencode.v19/ENCFF001RGR+spikes</td>\n",
        "      <td> 0</td>\n",
        "      <td>  10</td>\n",
        "      <td> 111807</td>\n",
        "      <td>   0</td>\n",
        "      <td> 111807</td>\n",
        "      <td>      0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>49</th>\n",
        "      <td> gencode.v19/ENCFF001RGR+spikes</td>\n",
        "      <td> 0</td>\n",
        "      <td>  36</td>\n",
        "      <td> 109103</td>\n",
        "      <td>   0</td>\n",
        "      <td>      0</td>\n",
        "      <td> 109103</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>50</th>\n",
        "      <td> gencode.v19/ENCFF001RGR+spikes</td>\n",
        "      <td> 0</td>\n",
        "      <td>  75</td>\n",
        "      <td> 105047</td>\n",
        "      <td>   0</td>\n",
        "      <td>      0</td>\n",
        "      <td> 105047</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>51</th>\n",
        "      <td> gencode.v19/ENCFF001RGR+spikes</td>\n",
        "      <td> 0</td>\n",
        "      <td> 100</td>\n",
        "      <td> 102447</td>\n",
        "      <td>   0</td>\n",
        "      <td>      0</td>\n",
        "      <td> 102447</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>52</th>\n",
        "      <td> gencode.v19/ENCFF001RGR+spikes</td>\n",
        "      <td> 1</td>\n",
        "      <td>  10</td>\n",
        "      <td> 111807</td>\n",
        "      <td>   0</td>\n",
        "      <td> 111807</td>\n",
        "      <td>      0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>53</th>\n",
        "      <td> gencode.v19/ENCFF001RGR+spikes</td>\n",
        "      <td> 1</td>\n",
        "      <td>  36</td>\n",
        "      <td> 109103</td>\n",
        "      <td>   3</td>\n",
        "      <td>      3</td>\n",
        "      <td> 109097</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>54</th>\n",
        "      <td> gencode.v19/ENCFF001RGR+spikes</td>\n",
        "      <td> 1</td>\n",
        "      <td>  75</td>\n",
        "      <td> 105047</td>\n",
        "      <td>   0</td>\n",
        "      <td>      0</td>\n",
        "      <td> 105047</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>55</th>\n",
        "      <td> gencode.v19/ENCFF001RGR+spikes</td>\n",
        "      <td> 1</td>\n",
        "      <td> 100</td>\n",
        "      <td> 102447</td>\n",
        "      <td>   0</td>\n",
        "      <td>      0</td>\n",
        "      <td> 102447</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>56</th>\n",
        "      <td> gencode.v19/ENCFF001RGR+spikes</td>\n",
        "      <td> 2</td>\n",
        "      <td>  10</td>\n",
        "      <td> 111807</td>\n",
        "      <td>   0</td>\n",
        "      <td> 111807</td>\n",
        "      <td>      0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>57</th>\n",
        "      <td> gencode.v19/ENCFF001RGR+spikes</td>\n",
        "      <td> 2</td>\n",
        "      <td>  36</td>\n",
        "      <td> 109103</td>\n",
        "      <td>  37</td>\n",
        "      <td>     36</td>\n",
        "      <td> 109030</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>58</th>\n",
        "      <td> gencode.v19/ENCFF001RGR+spikes</td>\n",
        "      <td> 2</td>\n",
        "      <td>  75</td>\n",
        "      <td> 105047</td>\n",
        "      <td>   0</td>\n",
        "      <td>      0</td>\n",
        "      <td> 105047</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>59</th>\n",
        "      <td> gencode.v19/ENCFF001RGR+spikes</td>\n",
        "      <td> 2</td>\n",
        "      <td> 100</td>\n",
        "      <td> 102447</td>\n",
        "      <td>   0</td>\n",
        "      <td>      0</td>\n",
        "      <td> 102447</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>60</th>\n",
        "      <td> gencode.v19/ENCFF001RGR+spikes</td>\n",
        "      <td> 3</td>\n",
        "      <td>  10</td>\n",
        "      <td> 111807</td>\n",
        "      <td>   0</td>\n",
        "      <td> 111807</td>\n",
        "      <td>      0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>61</th>\n",
        "      <td> gencode.v19/ENCFF001RGR+spikes</td>\n",
        "      <td> 3</td>\n",
        "      <td>  36</td>\n",
        "      <td> 109103</td>\n",
        "      <td> 106</td>\n",
        "      <td>    164</td>\n",
        "      <td> 108833</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>62</th>\n",
        "      <td> gencode.v19/ENCFF001RGR+spikes</td>\n",
        "      <td> 3</td>\n",
        "      <td>  75</td>\n",
        "      <td> 105047</td>\n",
        "      <td>   0</td>\n",
        "      <td>      0</td>\n",
        "      <td> 105047</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>63</th>\n",
        "      <td> gencode.v19/ENCFF001RGR+spikes</td>\n",
        "      <td> 3</td>\n",
        "      <td> 100</td>\n",
        "      <td> 102447</td>\n",
        "      <td>   0</td>\n",
        "      <td>      0</td>\n",
        "      <td> 102447</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>64</th>\n",
        "      <td> gencode.v19/ENCFF001RGS+spikes</td>\n",
        "      <td> 0</td>\n",
        "      <td>  10</td>\n",
        "      <td> 111807</td>\n",
        "      <td>   0</td>\n",
        "      <td> 111807</td>\n",
        "      <td>      0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>65</th>\n",
        "      <td> gencode.v19/ENCFF001RGS+spikes</td>\n",
        "      <td> 0</td>\n",
        "      <td>  36</td>\n",
        "      <td> 109103</td>\n",
        "      <td>   0</td>\n",
        "      <td>      0</td>\n",
        "      <td> 109103</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>66</th>\n",
        "      <td> gencode.v19/ENCFF001RGS+spikes</td>\n",
        "      <td> 0</td>\n",
        "      <td>  75</td>\n",
        "      <td> 105047</td>\n",
        "      <td>   0</td>\n",
        "      <td>      0</td>\n",
        "      <td> 105047</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>67</th>\n",
        "      <td> gencode.v19/ENCFF001RGS+spikes</td>\n",
        "      <td> 0</td>\n",
        "      <td> 100</td>\n",
        "      <td> 102447</td>\n",
        "      <td>   0</td>\n",
        "      <td>      0</td>\n",
        "      <td> 102447</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>68</th>\n",
        "      <td> gencode.v19/ENCFF001RGS+spikes</td>\n",
        "      <td> 1</td>\n",
        "      <td>  10</td>\n",
        "      <td> 111807</td>\n",
        "      <td>   0</td>\n",
        "      <td> 111807</td>\n",
        "      <td>      0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>69</th>\n",
        "      <td> gencode.v19/ENCFF001RGS+spikes</td>\n",
        "      <td> 1</td>\n",
        "      <td>  36</td>\n",
        "      <td> 109103</td>\n",
        "      <td>   3</td>\n",
        "      <td>      3</td>\n",
        "      <td> 109097</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>70</th>\n",
        "      <td> gencode.v19/ENCFF001RGS+spikes</td>\n",
        "      <td> 1</td>\n",
        "      <td>  75</td>\n",
        "      <td> 105047</td>\n",
        "      <td>   0</td>\n",
        "      <td>      0</td>\n",
        "      <td> 105047</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>71</th>\n",
        "      <td> gencode.v19/ENCFF001RGS+spikes</td>\n",
        "      <td> 1</td>\n",
        "      <td> 100</td>\n",
        "      <td> 102447</td>\n",
        "      <td>   0</td>\n",
        "      <td>      0</td>\n",
        "      <td> 102447</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>72</th>\n",
        "      <td> gencode.v19/ENCFF001RGS+spikes</td>\n",
        "      <td> 2</td>\n",
        "      <td>  10</td>\n",
        "      <td> 111807</td>\n",
        "      <td>   0</td>\n",
        "      <td> 111807</td>\n",
        "      <td>      0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>73</th>\n",
        "      <td> gencode.v19/ENCFF001RGS+spikes</td>\n",
        "      <td> 2</td>\n",
        "      <td>  36</td>\n",
        "      <td> 109103</td>\n",
        "      <td>  37</td>\n",
        "      <td>     36</td>\n",
        "      <td> 109030</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>74</th>\n",
        "      <td> gencode.v19/ENCFF001RGS+spikes</td>\n",
        "      <td> 2</td>\n",
        "      <td>  75</td>\n",
        "      <td> 105047</td>\n",
        "      <td>   0</td>\n",
        "      <td>      0</td>\n",
        "      <td> 105047</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>75</th>\n",
        "      <td> gencode.v19/ENCFF001RGS+spikes</td>\n",
        "      <td> 2</td>\n",
        "      <td> 100</td>\n",
        "      <td> 102447</td>\n",
        "      <td>   0</td>\n",
        "      <td>      0</td>\n",
        "      <td> 102447</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>76</th>\n",
        "      <td> gencode.v19/ENCFF001RGS+spikes</td>\n",
        "      <td> 3</td>\n",
        "      <td>  10</td>\n",
        "      <td> 111807</td>\n",
        "      <td>   0</td>\n",
        "      <td> 111807</td>\n",
        "      <td>      0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>77</th>\n",
        "      <td> gencode.v19/ENCFF001RGS+spikes</td>\n",
        "      <td> 3</td>\n",
        "      <td>  36</td>\n",
        "      <td> 109103</td>\n",
        "      <td> 106</td>\n",
        "      <td>    164</td>\n",
        "      <td> 108833</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>78</th>\n",
        "      <td> gencode.v19/ENCFF001RGS+spikes</td>\n",
        "      <td> 3</td>\n",
        "      <td>  75</td>\n",
        "      <td> 105047</td>\n",
        "      <td>   0</td>\n",
        "      <td>      0</td>\n",
        "      <td> 105047</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>79</th>\n",
        "      <td> gencode.v19/ENCFF001RGS+spikes</td>\n",
        "      <td> 3</td>\n",
        "      <td> 100</td>\n",
        "      <td> 102447</td>\n",
        "      <td>   0</td>\n",
        "      <td>      0</td>\n",
        "      <td> 102447</td>\n",
        "    </tr>\n",
        "  </tbody>\n",
        "</table>\n",
        "</div>"
       ],
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 13,
       "text": [
        "                             index  mismatch  read_length   total  hits  \\\n",
        "48  gencode.v19/ENCFF001RGR+spikes         0           10  111807     0   \n",
        "49  gencode.v19/ENCFF001RGR+spikes         0           36  109103     0   \n",
        "50  gencode.v19/ENCFF001RGR+spikes         0           75  105047     0   \n",
        "51  gencode.v19/ENCFF001RGR+spikes         0          100  102447     0   \n",
        "52  gencode.v19/ENCFF001RGR+spikes         1           10  111807     0   \n",
        "53  gencode.v19/ENCFF001RGR+spikes         1           36  109103     3   \n",
        "54  gencode.v19/ENCFF001RGR+spikes         1           75  105047     0   \n",
        "55  gencode.v19/ENCFF001RGR+spikes         1          100  102447     0   \n",
        "56  gencode.v19/ENCFF001RGR+spikes         2           10  111807     0   \n",
        "57  gencode.v19/ENCFF001RGR+spikes         2           36  109103    37   \n",
        "58  gencode.v19/ENCFF001RGR+spikes         2           75  105047     0   \n",
        "59  gencode.v19/ENCFF001RGR+spikes         2          100  102447     0   \n",
        "60  gencode.v19/ENCFF001RGR+spikes         3           10  111807     0   \n",
        "61  gencode.v19/ENCFF001RGR+spikes         3           36  109103   106   \n",
        "62  gencode.v19/ENCFF001RGR+spikes         3           75  105047     0   \n",
        "63  gencode.v19/ENCFF001RGR+spikes         3          100  102447     0   \n",
        "64  gencode.v19/ENCFF001RGS+spikes         0           10  111807     0   \n",
        "65  gencode.v19/ENCFF001RGS+spikes         0           36  109103     0   \n",
        "66  gencode.v19/ENCFF001RGS+spikes         0           75  105047     0   \n",
        "67  gencode.v19/ENCFF001RGS+spikes         0          100  102447     0   \n",
        "68  gencode.v19/ENCFF001RGS+spikes         1           10  111807     0   \n",
        "69  gencode.v19/ENCFF001RGS+spikes         1           36  109103     3   \n",
        "70  gencode.v19/ENCFF001RGS+spikes         1           75  105047     0   \n",
        "71  gencode.v19/ENCFF001RGS+spikes         1          100  102447     0   \n",
        "72  gencode.v19/ENCFF001RGS+spikes         2           10  111807     0   \n",
        "73  gencode.v19/ENCFF001RGS+spikes         2           36  109103    37   \n",
        "74  gencode.v19/ENCFF001RGS+spikes         2           75  105047     0   \n",
        "75  gencode.v19/ENCFF001RGS+spikes         2          100  102447     0   \n",
        "76  gencode.v19/ENCFF001RGS+spikes         3           10  111807     0   \n",
        "77  gencode.v19/ENCFF001RGS+spikes         3           36  109103   106   \n",
        "78  gencode.v19/ENCFF001RGS+spikes         3           75  105047     0   \n",
        "79  gencode.v19/ENCFF001RGS+spikes         3          100  102447     0   \n",
        "\n",
        "     multi  failed  \n",
        "48  111807       0  \n",
        "49       0  109103  \n",
        "50       0  105047  \n",
        "51       0  102447  \n",
        "52  111807       0  \n",
        "53       3  109097  \n",
        "54       0  105047  \n",
        "55       0  102447  \n",
        "56  111807       0  \n",
        "57      36  109030  \n",
        "58       0  105047  \n",
        "59       0  102447  \n",
        "60  111807       0  \n",
        "61     164  108833  \n",
        "62       0  105047  \n",
        "63       0  102447  \n",
        "64  111807       0  \n",
        "65       0  109103  \n",
        "66       0  105047  \n",
        "67       0  102447  \n",
        "68  111807       0  \n",
        "69       3  109097  \n",
        "70       0  105047  \n",
        "71       0  102447  \n",
        "72  111807       0  \n",
        "73      36  109030  \n",
        "74       0  105047  \n",
        "75       0  102447  \n",
        "76  111807       0  \n",
        "77     164  108833  \n",
        "78       0  105047  \n",
        "79       0  102447  "
       ]
      }
     ],
     "prompt_number": 13
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "Unsurprisingly 10 base pair sequences match all over the place. This suggests that setting the suppress multi-reads is a reasonable way to see if a read lands in more than one place. However. Its really rather odd that there's window-length reads that don't map correctly.\n",
      "\n",
      "That seems odd $--$ since we saved the repeat sequence lets look at one of them."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "! cat 0-mm/unmapped-spikes-36.fa"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        ">ERCC-00138_[368:404)\r\n",
        "ATTAGGGAAAATGGTGGAATTGTTGATAAAGTTTTY\r\n",
        ">ERCC-00138_[369:405)\r\n",
        "TTAGGGAAAATGGTGGAATTGTTGATAAAGTTTTYG\r\n",
        ">ERCC-00138_[370:406)\r\n",
        "TAGGGAAAATGGTGGAATTGTTGATAAAGTTTTYGT\r\n",
        ">ERCC-00138_[371:407)\r\n",
        "AGGGAAAATGGTGGAATTGTTGATAAAGTTTTYGTT\r\n",
        ">ERCC-00138_[372:408)\r\n",
        "GGGAAAATGGTGGAATTGTTGATAAAGTTTTYGTTG\r\n",
        ">ERCC-00138_[373:409)\r\n",
        "GGAAAATGGTGGAATTGTTGATAAAGTTTTYGTTGT\r\n",
        ">ERCC-00138_[374:410)\r\n",
        "GAAAATGGTGGAATTGTTGATAAAGTTTTYGTTGTT\r\n",
        ">ERCC-00138_[375:411)\r\n",
        "AAAATGGTGGAATTGTTGATAAAGTTTTYGTTGTTG\r\n",
        ">ERCC-00138_[376:412)\r\n",
        "AAATGGTGGAATTGTTGATAAAGTTTTYGTTGTTGT\r\n",
        ">ERCC-00138_[377:413)\r\n",
        "AATGGTGGAATTGTTGATAAAGTTTTYGTTGTTGTT\r\n",
        ">ERCC-00138_[378:414)\r\n",
        "ATGGTGGAATTGTTGATAAAGTTTTYGTTGTTGTTG\r\n",
        ">ERCC-00138_[379:415)\r\n",
        "TGGTGGAATTGTTGATAAAGTTTTYGTTGTTGTTGA\r\n",
        ">ERCC-00138_[380:416)\r\n",
        "GGTGGAATTGTTGATAAAGTTTTYGTTGTTGTTGAT\r\n",
        ">ERCC-00138_[381:417)\r\n",
        "GTGGAATTGTTGATAAAGTTTTYGTTGTTGTTGATA\r\n",
        ">ERCC-00138_[382:418)\r\n",
        "TGGAATTGTTGATAAAGTTTTYGTTGTTGTTGATAG\r\n",
        ">ERCC-00138_[383:419)\r\n",
        "GGAATTGTTGATAAAGTTTTYGTTGTTGTTGATAGG\r\n",
        ">ERCC-00138_[384:420)\r\n",
        "GAATTGTTGATAAAGTTTTYGTTGTTGTTGATAGGT\r\n",
        ">ERCC-00138_[385:421)\r\n",
        "AATTGTTGATAAAGTTTTYGTTGTTGTTGATAGGTT\r\n",
        ">ERCC-00138_[386:422)\r\n",
        "ATTGTTGATAAAGTTTTYGTTGTTGTTGATAGGTTA\r\n",
        ">ERCC-00138_[387:423)\r\n",
        "TTGTTGATAAAGTTTTYGTTGTTGTTGATAGGTTAG\r\n",
        ">ERCC-00138_[388:424)\r\n",
        "TGTTGATAAAGTTTTYGTTGTTGTTGATAGGTTAGA\r\n",
        ">ERCC-00138_[389:425)\r\n",
        "GTTGATAAAGTTTTYGTTGTTGTTGATAGGTTAGAA\r\n",
        ">ERCC-00138_[390:426)\r\n",
        "TTGATAAAGTTTTYGTTGTTGTTGATAGGTTAGAAG\r\n",
        ">ERCC-00138_[391:427)\r\n",
        "TGATAAAGTTTTYGTTGTTGTTGATAGGTTAGAAGG\r\n",
        ">ERCC-00138_[392:428)\r\n",
        "GATAAAGTTTTYGTTGTTGTTGATAGGTTAGAAGGA\r\n",
        ">ERCC-00138_[393:429)\r\n",
        "ATAAAGTTTTYGTTGTTGTTGATAGGTTAGAAGGAG\r\n",
        ">ERCC-00138_[394:430)\r\n",
        "TAAAGTTTTYGTTGTTGTTGATAGGTTAGAAGGAGC\r\n",
        ">ERCC-00138_[395:431)\r\n",
        "AAAGTTTTYGTTGTTGTTGATAGGTTAGAAGGAGCT\r\n",
        ">ERCC-00138_[396:432)\r\n",
        "AAGTTTTYGTTGTTGTTGATAGGTTAGAAGGAGCTA\r\n",
        ">ERCC-00138_[397:433)\r\n",
        "AGTTTTYGTTGTTGTTGATAGGTTAGAAGGAGCTAA\r\n",
        ">ERCC-00138_[398:434)\r\n",
        "GTTTTYGTTGTTGTTGATAGGTTAGAAGGAGCTAAA\r\n",
        ">ERCC-00138_[399:435)\r\n",
        "TTTTYGTTGTTGTTGATAGGTTAGAAGGAGCTAAAG\r\n",
        ">ERCC-00138_[400:436)\r\n",
        "TTTYGTTGTTGTTGATAGGTTAGAAGGAGCTAAAGA\r\n",
        ">ERCC-00138_[401:437)\r\n",
        "TTYGTTGTTGTTGATAGGTTAGAAGGAGCTAAAGAA\r\n",
        ">ERCC-00138_[402:438)\r\n",
        "TYGTTGTTGTTGATAGGTTAGAAGGAGCTAAAGAAA\r\n",
        ">ERCC-00138_[403:439)\r\n",
        "YGTTGTTGTTGATAGGTTAGAAGGAGCTAAAGAAAA\r\n"
       ]
      }
     ],
     "prompt_number": 14
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "Apparently ERCC-00138 has a Y in the sequence and bowtie doesn't know how to map that."
     ]
    },
    {
     "cell_type": "heading",
     "level": 2,
     "metadata": {},
     "source": [
      "Analyze Bowtie2"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "! /woldlab/castor/data00/proj/genome/programs/bowtie2-2.1.0/bowtie2 --version"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "/woldlab/castor/data00/proj/genome/programs/bowtie2-2.1.0/bowtie2-align version 2.1.0\r\n",
        "64-bit\r\n",
        "Built on do-dmxp-mac.win.ad.jhu.edu\r\n",
        "Tue Feb 26 13:34:02 EST 2013\r\n",
        "Compiler: gcc version 4.1.2 20080704 (Red Hat 4.1.2-54)\r\n",
        "Options: -O3 -m64 -msse2 -funroll-loops -g3 \r\n",
        "Sizeof {int, long, long long, void*, size_t, off_t}: {4, 8, 8, 8, 8, 8}\r\n"
       ]
      }
     ],
     "prompt_number": 15
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def parse_bowtie2(result):\n",
      "    parsed = {'total': 0,\n",
      "              'hits': 0,\n",
      "              'failed': 0,\n",
      "              'multi': 0}\n",
      "    \n",
      "    terms = [ ('(?P<total>[\\d]+) reads;', 'total'),\n",
      "              ('(?P<failed>[\\d]+).*aligned 0 times', 'failed'),\n",
      "              ('(?P<unique>[\\d]+).*aligned exactly 1 time', 'hits'),\n",
      "              ('(?P<multi>[\\d]+).*aligned >1 times', 'multi'),\n",
      "              ]\n",
      "    for line in result:\n",
      "        for pattern, variable in terms:\n",
      "            match = re.search(pattern, line)\n",
      "            if match:\n",
      "                parsed[variable] = int(match.group(1))\n",
      "    return parsed"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 16
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def bowtie2(index_pathname, read_length, verbose=False):\n",
      "    \"\"\"Run bowtie 2, using gravely lab settings \n",
      "    \"\"\"\n",
      "    path, index_name = os.path.split(index_pathname)\n",
      "    result = ! /woldlab/castor/data00/proj/genome/programs/bowtie2-2.1.0/bowtie2 \\\n",
      "                       -k 5 -f -p 4 \\\n",
      "                       -x {index_pathname} \\\n",
      "                       -U spikes-{read_length}.fa \\\n",
      "                       --un 0-mm/unmapped2-{index_name}-{read_length}.fa \\\n",
      "                       -S 0-mm/hits-{index_name}-{read_length}.sam\n",
      "    if verbose:\n",
      "        print(result)\n",
      "    return parse_bowtie2(result)\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 17
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def process_bowtie2(indexes, force=False, verbose=False):\n",
      "    # uncomment for testing.\n",
      "    #indexes = ['spikes']\n",
      "    cache_name = 'bowtie2'\n",
      "    if cache_name in cache and not force:\n",
      "        return cache[cache_name]\n",
      "    processed = []\n",
      "    for index in indexes:\n",
      "        for read_length in read_lengths:\n",
      "            properties = {\n",
      "               'index': index,\n",
      "               'read_length': read_length}\n",
      "            result = bowtie2(index, read_length, verbose)\n",
      "            properties.update(result)\n",
      "            processed.append(properties)\n",
      "    df =  pandas.DataFrame(processed, columns=['index', 'read_length', 'total', 'hits', 'multi', 'failed', ])\n",
      "    cache[cache_name] = df\n",
      "    return df"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 18
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "bowtie2_results = process_bowtie2(bowtie2_indexes, verbose=False)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 19
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "bowtie2_results"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "html": [
        "<div style=\"max-height:1000px;max-width:1500px;overflow:auto;\">\n",
        "<table border=\"1\" class=\"dataframe\">\n",
        "  <thead>\n",
        "    <tr style=\"text-align: right;\">\n",
        "      <th></th>\n",
        "      <th>index</th>\n",
        "      <th>read_length</th>\n",
        "      <th>total</th>\n",
        "      <th>hits</th>\n",
        "      <th>multi</th>\n",
        "      <th>failed</th>\n",
        "    </tr>\n",
        "  </thead>\n",
        "  <tbody>\n",
        "    <tr>\n",
        "      <th>0 </th>\n",
        "      <td>                    indexes2/spikes</td>\n",
        "      <td>  10</td>\n",
        "      <td> 111807</td>\n",
        "      <td>    366</td>\n",
        "      <td> 111441</td>\n",
        "      <td>      0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>1 </th>\n",
        "      <td>                    indexes2/spikes</td>\n",
        "      <td>  36</td>\n",
        "      <td> 109103</td>\n",
        "      <td> 109044</td>\n",
        "      <td>     51</td>\n",
        "      <td>      8</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>2 </th>\n",
        "      <td>                    indexes2/spikes</td>\n",
        "      <td>  75</td>\n",
        "      <td> 105047</td>\n",
        "      <td> 105047</td>\n",
        "      <td>      0</td>\n",
        "      <td>      0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>3 </th>\n",
        "      <td>                    indexes2/spikes</td>\n",
        "      <td> 100</td>\n",
        "      <td> 102447</td>\n",
        "      <td> 102447</td>\n",
        "      <td>      0</td>\n",
        "      <td>      0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>4 </th>\n",
        "      <td>        indexes2/ENCFF001RGR+spikes</td>\n",
        "      <td>  10</td>\n",
        "      <td> 111807</td>\n",
        "      <td>      0</td>\n",
        "      <td> 111807</td>\n",
        "      <td>      0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>5 </th>\n",
        "      <td>        indexes2/ENCFF001RGR+spikes</td>\n",
        "      <td>  36</td>\n",
        "      <td> 109103</td>\n",
        "      <td> 108852</td>\n",
        "      <td>    243</td>\n",
        "      <td>      8</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>6 </th>\n",
        "      <td>        indexes2/ENCFF001RGR+spikes</td>\n",
        "      <td>  75</td>\n",
        "      <td> 105047</td>\n",
        "      <td> 105047</td>\n",
        "      <td>      0</td>\n",
        "      <td>      0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>7 </th>\n",
        "      <td>        indexes2/ENCFF001RGR+spikes</td>\n",
        "      <td> 100</td>\n",
        "      <td> 102447</td>\n",
        "      <td> 102447</td>\n",
        "      <td>      0</td>\n",
        "      <td>      0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>8 </th>\n",
        "      <td>        indexes2/ENCFF001RGS+spikes</td>\n",
        "      <td>  10</td>\n",
        "      <td> 111807</td>\n",
        "      <td>      0</td>\n",
        "      <td> 111807</td>\n",
        "      <td>      0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>9 </th>\n",
        "      <td>        indexes2/ENCFF001RGS+spikes</td>\n",
        "      <td>  36</td>\n",
        "      <td> 109103</td>\n",
        "      <td> 108842</td>\n",
        "      <td>    253</td>\n",
        "      <td>      8</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>10</th>\n",
        "      <td>        indexes2/ENCFF001RGS+spikes</td>\n",
        "      <td>  75</td>\n",
        "      <td> 105047</td>\n",
        "      <td> 105047</td>\n",
        "      <td>      0</td>\n",
        "      <td>      0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>11</th>\n",
        "      <td>        indexes2/ENCFF001RGS+spikes</td>\n",
        "      <td> 100</td>\n",
        "      <td> 102447</td>\n",
        "      <td> 102447</td>\n",
        "      <td>      0</td>\n",
        "      <td>      0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>12</th>\n",
        "      <td> gencode.v19.bt2/ENCFF001RGR+spikes</td>\n",
        "      <td>  10</td>\n",
        "      <td> 111807</td>\n",
        "      <td>      0</td>\n",
        "      <td> 111807</td>\n",
        "      <td>      0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>13</th>\n",
        "      <td> gencode.v19.bt2/ENCFF001RGR+spikes</td>\n",
        "      <td>  36</td>\n",
        "      <td> 109103</td>\n",
        "      <td>     72</td>\n",
        "      <td>    103</td>\n",
        "      <td> 108928</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>14</th>\n",
        "      <td> gencode.v19.bt2/ENCFF001RGR+spikes</td>\n",
        "      <td>  75</td>\n",
        "      <td> 105047</td>\n",
        "      <td>      0</td>\n",
        "      <td>      0</td>\n",
        "      <td> 105047</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>15</th>\n",
        "      <td> gencode.v19.bt2/ENCFF001RGR+spikes</td>\n",
        "      <td> 100</td>\n",
        "      <td> 102447</td>\n",
        "      <td>      0</td>\n",
        "      <td>      0</td>\n",
        "      <td> 102447</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>16</th>\n",
        "      <td> gencode.v19.bt2/ENCFF001RGS+spikes</td>\n",
        "      <td>  10</td>\n",
        "      <td> 111807</td>\n",
        "      <td>      0</td>\n",
        "      <td> 111807</td>\n",
        "      <td>      0</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>17</th>\n",
        "      <td> gencode.v19.bt2/ENCFF001RGS+spikes</td>\n",
        "      <td>  36</td>\n",
        "      <td> 109103</td>\n",
        "      <td>     72</td>\n",
        "      <td>    106</td>\n",
        "      <td> 108925</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>18</th>\n",
        "      <td> gencode.v19.bt2/ENCFF001RGS+spikes</td>\n",
        "      <td>  75</td>\n",
        "      <td> 105047</td>\n",
        "      <td>      0</td>\n",
        "      <td>      0</td>\n",
        "      <td> 105047</td>\n",
        "    </tr>\n",
        "    <tr>\n",
        "      <th>19</th>\n",
        "      <td> gencode.v19.bt2/ENCFF001RGS+spikes</td>\n",
        "      <td> 100</td>\n",
        "      <td> 102447</td>\n",
        "      <td>      0</td>\n",
        "      <td>      0</td>\n",
        "      <td> 102447</td>\n",
        "    </tr>\n",
        "  </tbody>\n",
        "</table>\n",
        "</div>"
       ],
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 20,
       "text": [
        "                                 index  read_length   total    hits   multi  \\\n",
        "0                      indexes2/spikes           10  111807     366  111441   \n",
        "1                      indexes2/spikes           36  109103  109044      51   \n",
        "2                      indexes2/spikes           75  105047  105047       0   \n",
        "3                      indexes2/spikes          100  102447  102447       0   \n",
        "4          indexes2/ENCFF001RGR+spikes           10  111807       0  111807   \n",
        "5          indexes2/ENCFF001RGR+spikes           36  109103  108852     243   \n",
        "6          indexes2/ENCFF001RGR+spikes           75  105047  105047       0   \n",
        "7          indexes2/ENCFF001RGR+spikes          100  102447  102447       0   \n",
        "8          indexes2/ENCFF001RGS+spikes           10  111807       0  111807   \n",
        "9          indexes2/ENCFF001RGS+spikes           36  109103  108842     253   \n",
        "10         indexes2/ENCFF001RGS+spikes           75  105047  105047       0   \n",
        "11         indexes2/ENCFF001RGS+spikes          100  102447  102447       0   \n",
        "12  gencode.v19.bt2/ENCFF001RGR+spikes           10  111807       0  111807   \n",
        "13  gencode.v19.bt2/ENCFF001RGR+spikes           36  109103      72     103   \n",
        "14  gencode.v19.bt2/ENCFF001RGR+spikes           75  105047       0       0   \n",
        "15  gencode.v19.bt2/ENCFF001RGR+spikes          100  102447       0       0   \n",
        "16  gencode.v19.bt2/ENCFF001RGS+spikes           10  111807       0  111807   \n",
        "17  gencode.v19.bt2/ENCFF001RGS+spikes           36  109103      72     106   \n",
        "18  gencode.v19.bt2/ENCFF001RGS+spikes           75  105047       0       0   \n",
        "19  gencode.v19.bt2/ENCFF001RGS+spikes          100  102447       0       0   \n",
        "\n",
        "    failed  \n",
        "0        0  \n",
        "1        8  \n",
        "2        0  \n",
        "3        0  \n",
        "4        0  \n",
        "5        8  \n",
        "6        0  \n",
        "7        0  \n",
        "8        0  \n",
        "9        8  \n",
        "10       0  \n",
        "11       0  \n",
        "12       0  \n",
        "13  108928  \n",
        "14  105047  \n",
        "15  102447  \n",
        "16       0  \n",
        "17  108925  \n",
        "18  105047  \n",
        "19  102447  "
       ]
      }
     ],
     "prompt_number": 20
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "plot"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "ename": "NameError",
       "evalue": "name 'plot' is not defined",
       "output_type": "pyerr",
       "traceback": [
        "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m\n\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
        "\u001b[0;32m<ipython-input-21-88d627f14d84>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mplot\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
        "\u001b[0;31mNameError\u001b[0m: name 'plot' is not defined"
       ]
      }
     ],
     "prompt_number": 21
    },
    {
     "cell_type": "heading",
     "level": 2,
     "metadata": {},
     "source": [
      "Further Work"
     ]
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "* Need to align real data against spikes and make sure nothing accidentally hits."
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": []
    }
   ],
   "metadata": {}
  }
 ]
}